sw/host/opentitanlib/src/bootstrap/legacy.rs

```diff
@@ -0,0 +1,285 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use mundane::hash::{Digest, Hasher, Sha256};
+use std::time::Duration;
+use thiserror::Error;
+use zerocopy::AsBytes;
+
+use crate::bootstrap::{BootstrapOptions, UpdateProtocol};
+use crate::io::spi::{Target, Transfer};
+
+#[derive(AsBytes, Debug, Default)]
+#[repr(C)]
+struct FrameHeader {
+    hash: [u8; Frame::HASH_LEN],
+    frame_num: u32,
+    flash_offset: u32,
+}
+
+#[derive(AsBytes, Debug)]
+#[repr(C)]
+struct Frame {
+    header: FrameHeader,
+    data: [u8; Frame::DATA_LEN],
+}
+
+impl Default for Frame {
+    fn default() -> Self {
+        Frame {
+            header: Default::default(),
+            data: [0xff; Frame::DATA_LEN],
+        }
+    }
+}
+
+impl Frame {
+    const EOF: u32 = 0x8000_0000;
+    const FLASH_SECTOR_SIZE: usize = 2048;
+    const FLASH_BUFFER_SIZE: usize = 128;
+    const DATA_LEN: usize = 2048 - std::mem::size_of::<FrameHeader>();
+    const HASH_LEN: usize = 32;
+    const FLASH_BASE_ADDRESS: usize = 65536 * 8;
+
+    /// Computes the hash in the header.
+    fn header_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let frame = self.as_bytes();
+        let sha = Sha256::hash(&frame[Frame::HASH_LEN..]);
+        sha.bytes()
+    }
+
+    /// Computes the hash over the entire frame.
+    fn frame_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let sha = Sha256::hash(self.as_bytes());
+        let mut digest = sha.bytes();
+        // Touch up zeroes into ones, as that is what the old chips are doing.
+        for i in 0..digest.len() {
+            if digest[i] == 0 {
+                digest[i] = 1;
+            }
+        }
+        digest
+    }
+
+    /// Creates a sequence of frames based on a `payload` binary.
+    fn from_payload(payload: &[u8]) -> Vec<Frame> {
+        let mut frames = Vec::new();
+
+        let mut max_addr = (payload.len() + 3) & !3;
+        while max_addr > 0 && payload[max_addr - 4..max_addr] == [0xff, 0xff, 0xff, 0xff] {
+            max_addr -= 4;
+        }
+
+        let mut frameno = 0;
+        let mut addr = 0;
+        while addr < max_addr {
+            // Try skipping over 0xffffffff words.
+            let mut nonempty_addr = addr;
+            while nonempty_addr < max_addr {
+                if payload[nonempty_addr..nonempty_addr + 4] != [0xff, 0xff, 0xff, 0xff] {
+                    break;
+                }
+                nonempty_addr += 4;
+            }
+            if nonempty_addr >= max_addr {
+                // Can skip all the way to end; signal last frame.
+                frameno |= Self::EOF;
+            } else {
+                // Can skip to upto nonempty.
+                let skip_addr = nonempty_addr & !(Self::FLASH_SECTOR_SIZE - 1);
+                if skip_addr > addr {
+                    if addr == 0 {
```

Comment:
Feels like these two conditions should be mered?

Reply:
I have simplified the code in this section a bit, since I realized that we do not nave to consider the case that we find 0xFF all the way to max_addr, since max_addr is already pointing at a "trimmed" end.

I do not quite understand what you suggest that I should merge, though.